File: haskell/clock/src/Clock.hs
module Clock (addDelta, fromHourMin, toString) where

type Minutes = Int
type Hours   = Int
data Clock = Clock {

        hours   :: Hours
    ,   minutes :: Minutes

    } deriving (Eq, Ord)

instance Show Clock where
    show Clock { hours = h
               , minutes = m } =
        (pad . show) h ++ ":" ++ (pad . show) m
        where pad str = replicate (2 - length str) '0' ++ str


fromHourMin :: Hours -> Minutes -> Clock
fromHourMin hour mins =
    Clock { hours = (hour + (mins `div` 60)) `mod` 24
          , minutes = mins `mod` 60 }


toString :: Clock -> String
toString = show


addDelta :: Hours -> Minutes -> Clock -> Clock
addDelta hour mins clock =
    fromHourMin (hour + hours clock)
                (mins + minutes clock)

File: javascript/clock/clock.js
function mod(n, m) {
  return ((n % m) + m) % m;
}

export class Clock {

  constructor(hours=0, minutes=0) {
    this._minutes = 0;
    this._hours = 0;
    this.plus(hours * 60 + minutes);
  }

  toString() {
    return  String(this._hours).padStart(2, '0')
        +   ":"
        +   String(this._minutes).padStart(2, '0');
  }

  plus(minutes) {
    this._hours = mod(this._hours + Math.floor((this._minutes + minutes) / 60), 24);
    this._minutes = mod(this._minutes + minutes, 60);
    return this;
  }

  minus(minutes) {
    return this.plus(-minutes);
  }

  equals(other) {
    return this._minutes === other._minutes
          &&  this._hours === other._hours;
  }
}
